File: scripts/validate.ts
import { parse as parseToml } from "@std/toml";
import { parseAll as parseYaml } from "@std/yaml";

const ROOT = new URL("../", import.meta.url);
const MARKDOWN_LINK = /!?\[[^\]]*]\(([^)]+)\)/g;
const STRUCTURED_LOADERS: Record<string, (text: string) => unknown> = {
  ".json": JSON.parse,
  ".toml": parseToml,
  ".yaml": parseYaml,
  ".yml": parseYaml,
};

async function* repositoryFiles(directory = ROOT): AsyncGenerator<URL> {
  for await (const entry of Deno.readDir(directory)) {
    if (entry.name === ".git") continue;

    const suffix = entry.isDirectory ? "/" : "";
    const url = new URL(
      `${encodeURIComponent(entry.name)}${suffix}`,
      directory,
    );
    if (entry.isDirectory) {
      yield* repositoryFiles(url);
    } else if (entry.isFile) {
      yield url;
    }
  }
}

function relativePath(url: URL): string {
  return decodeURIComponent(url.pathname.slice(ROOT.pathname.length));
}

function extension(url: URL): string {
  return url.pathname.match(/\.[^.\/]+$/)?.[0].toLowerCase() ?? "";
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

async function validateStructuredFiles(errors: string[]): Promise<number> {
  let checked = 0;

  for await (const file of repositoryFiles()) {
    const loader = STRUCTURED_LOADERS[extension(file)];
    if (!loader) continue;

    checked += 1;
    try {
      loader(await Deno.readTextFile(file));
    } catch (error) {
      errors.push(`${relativePath(file)}: ${errorMessage(error)}`);
    }
  }

  return checked;
}

function localLinkUrl(markdown: URL, rawTarget: string): URL | undefined {
  let target = rawTarget.trim();
  if (target.startsWith("<") && target.endsWith(">")) {
    target = target.slice(1, -1);
  } else {
    target = target.split(/\s+/, 1)[0];
  }

  if (!target || target.startsWith("#")) return;

  const url = new URL(target, markdown);
  return url.protocol === "file:" ? url : undefined;
}

async function exists(url: URL): Promise<boolean> {
  try {
    await Deno.stat(url);
    return true;
  } catch (error) {
    if (error instanceof Deno.errors.NotFound) return false;
    throw error;
  }
}

async function validateMarkdownLinks(errors: string[]): Promise<number> {
  let checked = 0;

  for await (const markdown of repositoryFiles()) {
    if (extension(markdown) !== ".md") continue;

    const text = await Deno.readTextFile(markdown);
    for (const match of text.matchAll(MARKDOWN_LINK)) {
      const target = localLinkUrl(markdown, match[1]);
      if (!target) continue;

      checked += 1;
      if (!(await exists(target))) {
        errors.push(
          `${relativePath(markdown)}: missing local link ${
            JSON.stringify(match[1])
          }`,
        );
      }
    }
  }

  return checked;
}

async function validateAgentInstructions(errors: string[]): Promise<void> {
  const agents = new URL("AGENTS.md", ROOT);
  const claude = new URL("CLAUDE.md", ROOT);

  if (!(await exists(agents))) {
    errors.push("AGENTS.md: canonical agent instructions are missing");
  } else if ((await Deno.stat(agents)).size > 32 * 1024) {
    errors.push("AGENTS.md: exceeds Codex's default 32 KiB instruction limit");
  }

  if (
    !(await exists(claude)) ||
    (await Deno.readTextFile(claude)).trim() !== "@AGENTS.md"
  ) {
    errors.push("CLAUDE.md: expected the single adapter line '@AGENTS.md'");
  }
}

async function main(): Promise<void> {
  const errors: string[] = [];
  const structured = await validateStructuredFiles(errors);
  const links = await validateMarkdownLinks(errors);
  await validateAgentInstructions(errors);

  if (errors.length > 0) {
    console.error("Validation failed:");
    for (const error of errors) console.error(`  - ${error}`);
    Deno.exit(1);
  }

  console.log(
    `Validated ${structured} structured files and ${links} local links.`,
  );
}

await main();

File: scripts/validate.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.11"
# dependencies = ["pyyaml"]
# ///

"""Validate structured configuration and repository-local Markdown links."""

from __future__ import annotations

import json
import re
import sys
import tomllib
from pathlib import Path
from urllib.parse import unquote, urlsplit

import yaml

ROOT = Path(__file__).resolve().parent.parent
MARKDOWN_LINK = re.compile(r"!?\[[^\]]*]\(([^)]+)\)")


def repository_files(suffixes: set[str]) -> list[Path]:
    return sorted(
        path
        for path in ROOT.rglob("*")
        if path.is_file()
        and path.suffix.lower() in suffixes
        and ".git" not in path.parts
    )


def validate_structured_files(errors: list[str]) -> dict[str, int]:
    loaders = {
        ".json": lambda text: json.loads(text),
        ".toml": lambda text: tomllib.loads(text),
        ".yaml": lambda text: list(yaml.safe_load_all(text)),
        ".yml": lambda text: list(yaml.safe_load_all(text)),
    }
    counts = {suffix: 0 for suffix in loaders}

    for path in repository_files(set(loaders)):
        suffix = path.suffix.lower()
        counts[suffix] += 1
        try:
            loaders[suffix](path.read_text(encoding="utf-8"))
        except Exception as error:
            errors.append(f"{path.relative_to(ROOT)}: {error}")

    return counts


def local_link_path(markdown: Path, raw_target: str) -> Path | None:
    target = raw_target.strip()
    if target.startswith("<") and target.endswith(">"):
        target = target[1:-1]
    else:
        target = target.split(maxsplit=1)[0]

    parsed = urlsplit(target)
    if parsed.scheme or parsed.netloc or target.startswith("#"):
        return None

    path = unquote(parsed.path)
    if not path:
        return None
    return (markdown.parent / path).resolve()


def validate_markdown_links(errors: list[str]) -> int:
    checked = 0
    for markdown in repository_files({".md"}):
        text = markdown.read_text(encoding="utf-8")
        for match in MARKDOWN_LINK.finditer(text):
            target = local_link_path(markdown, match.group(1))
            if target is None:
                continue
            checked += 1
            if not target.exists():
                errors.append(
                    f"{markdown.relative_to(ROOT)}: missing local link "
                    f"{match.group(1)!r}"
                )
    return checked


def validate_agent_instructions(errors: list[str]) -> None:
    agents = ROOT / "AGENTS.md"
    claude = ROOT / "CLAUDE.md"

    if not agents.is_file():
        errors.append("AGENTS.md: canonical agent instructions are missing")
    elif agents.stat().st_size > 32 * 1024:
        errors.append("AGENTS.md: exceeds Codex's default 32 KiB instruction limit")

    if not claude.is_file() or claude.read_text(encoding="utf-8").strip() != "@AGENTS.md":
        errors.append("CLAUDE.md: expected the single adapter line '@AGENTS.md'")


def main() -> int:
    errors: list[str] = []
    counts = validate_structured_files(errors)
    links = validate_markdown_links(errors)
    validate_agent_instructions(errors)

    if errors:
        print("Validation failed:", file=sys.stderr)
        for error in errors:
            print(f"  - {error}", file=sys.stderr)
        return 1

    structured = sum(counts.values())
    print(f"Validated {structured} structured files and {links} local links.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
